fetch/src/main/java/com/indix/gocd/s3fetch/FetchExecutor.java

```diff
@@ -14,37 +16,82 @@
+            if (!validationResult.isSuccessful()) {
+                logger.error(String.format("s3 fetch configuration is invalid: %s", validationResult.getMessages().toString()));
+                return ExecutionResult.failure(validationResult.getMessages().toString());
+            }
+            final S3ArtifactStore store = s3ArtifactStore(fetchConfig);
 
-        String artifactPathOnS3 = fetchConfig.getArtifactsLocationTemplate();
-        context.console().printLine(String.format("Getting artifacts from %s", store.pathString(artifactPathOnS3)));
-        String destination = String.format("%s/%s", context.workingDir(), config.getValue(FetchTask.DESTINATION));
-        setupDestinationDirectory(destination);
+            String artifactPathOnS3 = fetchConfig.getArtifactsLocationTemplate();
+            context.console().printLine(String.format("Getting artifacts from %s", store.pathString(artifactPathOnS3)));
+            String destination = String.format("%s/%s", context.workingDir(), config.getValue(FetchTask.DESTINATION));
+            setupDestinationDirectory(destination);
 
-        try {
-            store.getPrefix(artifactPathOnS3, destination);
-        } catch (Exception e) {
-            String message = String.format("Failure while downloading artifacts - %s", e.getMessage());
-            logger.error(message, e);
+            try {
+                store.getPrefix(artifactPathOnS3, destination);
+            } catch (Exception e) {
+                String message = String.format("Failure while downloading artifacts - %s", e.getMessage());
+                logger.error(message, e);
+                return ExecutionResult.failure(message, e);
+            }
+
+            ExecutionResult zipArchivesExecutionResult = ExtractAndCleanupZipArchivesIfPresent(destination);
+
+            if (!zipArchivesExecutionResult.isSuccessful()) {
+                return zipArchivesExecutionResult;
+            }
+            else {
+                return ExecutionResult.success("Fetched all artifacts");
+            }
+        }
+        catch (Exception e) {
+            String message = String.format("Error during fetch from s3: %s", e.getMessage());
+            logger.error(message);
             return ExecutionResult.failure(message, e);
         }
+    }
 
-        return ExecutionResult.success("Fetched all artifacts");
+    private ExecutionResult ExtractAndCleanupZipArchivesIfPresent(String destination)
+    {
+        List<File> files = (List<File>) FileUtils.listFiles(new File(destination), new String[]{"zip"}, true);
+        for (File zipFile : files) {
+            if (zipFile.getName().endsWith("artifacts.zip")) {
+                try {
+                    logger.debug(String.format("Artifact is archive.zip: un-compressing %s into %s",
+                            zipFile.getAbsolutePath(), zipFile.getParent()));
+                    zipArchiveManager.extractArchive(zipFile.getAbsolutePath(), zipFile.getParent());
```

Comment:
But, shouldn't we cleanup on failure? I don't know if the library that's being used takes care of cleaning up out dirs on archive corruption, etc. Let's please handle it to ensure we don't have rogue archives causing problems.